src/parsing/chunked_reader.rs

```diff
@@ -34,95 +32,64 @@ where
     pub fn new(reader: BufReader<R>) -> ChunkedReader<R> {
         ChunkedReader {
             inner: reader,
-            is_expecting_chunk: true,
-            read: 0,
-            length: 0,
-            line: Vec::new(),
+            buffer: Vec::new(),
+            consumed: 0,
+            reached_eof: false,
         }
     }
 
-    #[inline]
-    fn remaining(&self) -> u64 {
-        self.length - self.read
-    }
-
-    #[inline]
-    fn read_line(&mut self) -> io::Result<usize> {
-        buffers::read_line(&mut self.inner, &mut self.line)
-    }
-
-    fn read_chunk_size(&mut self) -> io::Result<u64> {
-        self.read_line()?;
-        if self.line.is_empty() {
+    fn read_chunk_size(&mut self) -> io::Result<usize> {
+        buffers::read_line(&mut self.inner, &mut self.buffer)?;
+        if self.buffer.is_empty() {
             return Err(io::ErrorKind::UnexpectedEof.into());
         }
-        parse_chunk_size(&self.line)
-    }
-
-    fn read_empty_line(&mut self) -> io::Result<()> {
-        let n = self.read_line()?;
-        if n == 0 || !self.line.is_empty() {
-            Err(InvalidResponseKind::Chunk.into())
-        } else {
-            Ok(())
-        }
+        parse_chunk_size(&self.buffer)
     }
 }
 
-impl<R> Read for ChunkedReader<R>
+impl<R> BufRead for ChunkedReader<R>
 where
     R: Read,
 {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        if self.is_expecting_chunk {
-            debug!("waiting, parsing new chunk size");
-            // If we're waiting for a new chunk, we read a line and parse the number as hexadecimal.
-            self.read = 0;
-            self.length = self.read_chunk_size()?;
-            // If the chunk's length is 0, we've received the EOF chunk.
-            if self.length == 0 {
-                debug!("received EOF chunk");
-                // Read CRLF
-                self.read_empty_line()?;
-            }
-            self.is_expecting_chunk = false;
-        }
+    fn fill_buf(&mut self) -> io::Result<&[u8]> {
+        if self.buffer.len() == self.consumed && !self.reached_eof {
+            let chunk_size = self.read_chunk_size()?;
 
-        // If we have a length of 0 for a chunk, we've reached EOF.
-        if self.length == 0 {
-            return Ok(0);
-        }
+            self.buffer.resize(chunk_size + 2, 0);
+            self.inner.read_exact(&mut self.buffer)?;
 
-        // We read the smallest amount between the given buffer's length and the remaining bytes' length.
-        let count = cmp::min(self.remaining(), buf.len() as u64) as usize;
+            if !self.buffer.ends_with(b"\r\n") {
```

Comment:
The current way accepts both `\n` and `\r\n`. I like the saying "be conservative in what you do, be liberal in what you accept from others". I don't think it's easy to accept both ways with a read_exact though.

Reply:
Yes, that is certainly a change in accepted data. But I think it can be easily fixing with a bit of ceremony as the underlying reader is already buffered, i.e. I can `read_exact` only `chunk_size + 1` and then read another character `\r\n` if the last one was `\r`.

Reply:
Pushed a fixed version factoring out a `buffers::read_line_ending` helper.